misk-hibernate/src/main/kotlin/misk/vitess/StartVitessService.kt

```diff
@@ -313,14 +313,32 @@ class LogContainerResultCallback : ResultCallbackTemplate<LogContainerResultCall
  * will be stopped and removed. Otherwise (if the container was started by a different process), it
  * will be left running.
  */
-class StartVitessService(
-  private val qualifier: KClass<out Annotation>,
-  private val environment: Environment,
+class StartVitessService : AbstractIdleService, DependentService {
+
+  private val qualifier: KClass<out Annotation>
+  private val environment: Environment
   private val config: DataSourceConfig
-) : AbstractIdleService(), DependentService {
 
-  override val consumedKeys: Set<Key<*>> = setOf()
-  override val producedKeys: Set<Key<*>> = setOf(Key.get(StartVitessService::class.java))
+  constructor(
+    qualifier: KClass<out Annotation>,
+    environment: Environment,
+    config: DataSourceConfig
+  ) : super() {
+    this.qualifier = qualifier
+    this.environment = environment
+    this.config = config
+    this.consumedKeys = setOf()
+    this.producedKeys = setOf(Key.get(StartVitessService::class.java))
+
+    if (config.type == DataSourceType.VITESS) {
+      // Need to pull the docker image here,
+      // can't do it during service start up because it's too slow
+      StartVitessService.pullImage()
```

Comment:
I have to confess that I don't really like doing things like this in a constructor, but neither do I like doing it in a class initializer either so I guess this doesn't really make things worse either.

Reply:
Your other option is to do it in the `configure()` function of the vitess module. That might be slightly better?